### PR TITLE
Add SwaV ImageNet benchmark

### PR DESCRIPTION
**Pull request by philippmwirth**
# Add SwaV ImageNet benchmark

Closes https://github.com/lightly-ai/lightly/issues/1230. 
* Add Code for SwAV ImageNet benchmark.
* Update tables with results.
* Fix imports by adding `lightly.utils.benchmarking` to `setup.py`.
* Fix instructions to run benchmarks.

## Results
Tensorboard link: https://tensorboard.dev/experiment/Ipx4Oxl5Qkqm5Sl5kWyKKg/#scalars
```
Imagenet:
    knn val_top1: 0.49526000022888184
    knn val_top5: 0.7857199907302856
    max linear val_top1: 0.671999990940094
    max linear val_top5: 0.8811799883842468
    max finetune val_top1: 0.7540599703788757
    max finetune val_top5: 0.9268800020217896
```
Training was 100 epochs with batch size 256 (=2x128) to match the experiment settings of the SimCLR benchmarks. Unfortunately, the SwAV paper only has numbers for 200 epochs with batch size 256:
image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -287,6 +287,7 @@ tuned for maximum accuracy. For detailed results and more info about the benchma
 | Model       | Backbone | Batch Size | Epochs | Linear Top1 | Linear Top1 Online | KNN Top1 | Tensorboard | Checkpoint |
 |-------------|----------|------------|--------|-------------|--------------------|----------|-------------|------------|
 | SimCLR      | Res50    |        256 |    100 |        63.2 |               63.1 |     44.9 |      [link](https://tensorboard.dev/experiment/JwNs9E02TeeQkS7aljh8dA) |       [link](https://lightly-ssl-checkpoints.s3.amazonaws.com/imagenet_resnet50_simclr_2023-05-04_09-02-54/pretrain/version_0/checkpoints/epoch%3D99-step%3D500400.ckpt) |
+| SwAV        | Res50    |        256 |    100 |        67.2 |               62.1 |     49.5 |      [link](https://tensorboard.dev/experiment/Ipx4Oxl5Qkqm5Sl5kWyKKg) |       [link](https://lightly-ssl-checkpoints.s3.amazonaws.com/imagenet_resnet50_swav_2023-05-25_08-29-14/pretrain/version_0/checkpoints/epoch%3D99-step%3D500400.ckpt)
```
Comment (guarin) [inline README.md:290]: Let's add finetune top1 and drop the online linear top1 column instead.

File: benchmarks/imagenet/resnet50/swav.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    def training_step(
+        self, batch: Tuple[List[Tensor], Tensor, List[str]], batch_idx: int
+    ) -> Tensor:
+        # Normalize the prototypes so they are on the unit sphere.
+        self.prototypes.normalize()
+
+        # The multi-crop dataloader returns a list of image crops where the
+        # first few items are high resolution crops and the rest are low
+        # resolution crops.
+        multi_crops, targets, _ = batch
```
Comment (guarin) [inline benchmarks/imagenet/resnet50/swav.py:65]:
```suggestion
        multi_crops, targets = batch[0], batch[1]
```

This way it is also compatible with torch datasets.

File: benchmarks/imagenet/resnet50/swav.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+    def training_step(
+        self, batch: Tuple[List[Tensor], Tensor, List[str]], batch_idx: int
+    ) -> Tensor:
+        # Normalize the prototypes so they are on the unit sphere.
+        self.prototypes.normalize()
+
+        # The multi-crop dataloader returns a list of image crops where the
+        # first few items are high resolution crops and the rest are low
+        # resolution crops.
```
Comment (guarin) [inline benchmarks/imagenet/resnet50/swav.py:64]:
```suggestion
        # The dataloader returns a list of image crops where the
        # first few items are high resolution crops and the rest are low
        # resolution crops.
```

File: docs/source/getting_started/benchmarks.rst [hunk 4]
```diff
@@ -25,7 +25,7 @@ the code at `benchmarks/imagenet/resnet50 <https://github.com/lightly-ai/lightly
   :widths: 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20
 
   "SimCLR", "Res50", "256", "100", "63.2", "85.3", "63.1", "85.2", "44.9", "74.2", "`link <https://tensorboard.dev/experiment/JwNs9E02TeeQkS7aljh8dA>`_", "`link <https://lightly-ssl-checkpoints.s3.amazonaws.com/imagenet_resnet50_simclr_2023-05-04_09-02-54/pretrain/version_0/checkpoints/epoch%3D99-step%3D500400.ckpt>`_"
-
+  "SwAV", "Res50", "256", "100", "67.2", "88.1", "62.1", "85.4", "49.5", "78.6", "`link <https://tensorboard.dev/experiment/Ipx4Oxl5Qkqm5Sl5kWyKKg>`_", "`link <https://lightly-ssl-checkpoints.s3.amazonaws.com/imagenet_resnet50_swav_2023-05-25_08-29-14/pretrain/version_0/checkpoints/epoch%3D99-step%3D500400.ckpt>`_"
```
Comment (guarin) [inline docs/source/getting_started/benchmarks.rst:28]: Let's also add finetune results.

**Review by guarin — CHANGES_REQUESTED** [review]
Awesome!

The results look good to me. I guess we should add a note to our benchmark results that we use SimCLR evaluation settings for all benchmark as they differ from SwAV.

Main differences:
* lr: SimCLR=0.1, SwAV=0.3
* epochs: SimCLR=90, SwAV=100
* weight decay: SimCLR=0, SwAV=1e-6

This is probably also the reason why the results are a bit lower than expected. Note that batch size should only have a minimal impact (from [vissl](https://github.com/facebookresearch/vissl/blob/main/MODEL_ZOO.md#swav)):
```
RN50 - 200 epochs - 2x224+6x96 - 256 batch-size: 73.07 lin. top1
RN50 - 200 epochs - 2x224+6x96 - 4096 batch-size: 73.85 lin. top1
RN50 - 100 epochs - 2x224+6x96 - 4096 batch-size: 71.99 lin. top1
```

So with SwAV linear eval settings I would expect ~71.5 lin. top1

---

What is a bit surprising to me are the KNN results. The paper reports much higher accuracy (table 11 in the appendix):

Screen Shot 2023-05-30 at 09 44 28

Not sure where the difference comes from. Maybe the implementation differs, but I wasn't able to find the knn evaluation code in the original repo.

File: docs/source/getting_started/benchmarks.rst (hunk 4 above)
Comment (philippmwirth) [reply docs/source/getting_started/benchmarks.rst:28]: Can you provide the ones for simclr?

Comment (guarin) [reply docs/source/getting_started/benchmarks.rst:28]: The SimCLR benchmarks didn't include finetuning. I would just leave it empty for now.

**Review by guarin — APPROVED** [review]
LGTM!
